internal/core/metadata/rest_addressable_test.go

```diff
@@ -93,7 +94,11 @@ func TestGetAllAddressables(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 
 			rr := httptest.NewRecorder()
-			restGetAllAddressables(rr, logger.NewMockClient(), tt.dbMock)
+			restGetAllAddressables(
+				rr,
+				logger.NewMockClient(),
+				tt.dbMock,
+				errorconcept.NewErrorHandler(logger.NewMockClient()))
```

Comment:
Commented once; applies elsewhere.

Should `errorconcept.ErrorHandler` be an actual instance or a mock?

Reply:
Good catch! @akramtexas  and I were looking at getting this changed to used a mock and we ran into a few hurdles. The issue is the expected side effect in the `ErrorHandler` where the HTTP status code is being determined and the written to the HTTP response:

https://github.com/edgexfoundry/edgex-go/blob/21750bd69222753f69a7a217e480b07ca6bb6dd6/internal/pkg/errorconcept/handler.go#L44-L49

The issue is the tests are validating the the response has been passed to the `http.ResponseWriter` and providing a simple mock will not perform that operation. This lead to us creating an anonymous function to be executed when the mock was invoked that would pass the response to the `http.ResponseWriter`. This approach is not really desirable as it invalidates the test since we are just writing the "hard-coded" status code value to the `http.ResponseWriter` rather than letting the REST handler process the request and populate the response naturally.

I think these issues are based on that the tests are not structured properly since the addition of the `ErrorHandler`. The tests now are testing two things:

1. The REST handler's functionality
2. The `ErrorHandler` functionality

I think these should be split up into two different test suites. The tests which validate the REST handlers can verify the functionality that is explicitly performed by the REST handler --like ensuring the proper `Content-Type` is written to the response-- and for tests with expected errors we can simply verify that the `ErrorHandler` mock was called with the proper parameters rather than verifying the status code. We could then test the propagation of the status code from the `ErrorHandler` to the `http.ResponseWriter` within the tests specifically for the `ErrorHandler`. 

I hope this makes sense. Any thoughts?

Reply:
- Yep, I agree, @AnthonyMBonafide, that this is definitely a good catch by @michaelestrin in his observation (_"Should errorconcept.ErrorHandler be an actual instance or a mock?"_). 
- It led me, during implementing the approach of mocking `errorconcept.ErrorHandler`, to investigate the root cause of the following scenario wherein I would get test failures due to mismatches on the expected HTTP response code, for example:
```
--- FAIL: TestGetAllAddressables/Error_Limit_Exceeded (0.00s)
    rest_addressable_test.go:133: status code mismatch -- expected 413 got 200
--- FAIL: TestGetAllAddressables/Error_Unknown (0.00s)
    rest_addressable_test.go:133: status code mismatch -- expected 500 got 200
```
- And which is when I had reached out to bounce ideas off you, get some ideas. Thanks again for making the time.
- Exactly, tracing back to the fact that the response code is not explicitly written by `http.ResponseWriter` in the following three relevant functions for ErrorHandler:
```
- Handle(w http.ResponseWriter, err error, ec ErrorConceptType)
- HandleManyVariants(w http.ResponseWriter, err error, allowableErrors []ErrorConceptType, defaultError ErrorConceptType)
- HandleOneVariant(w http.ResponseWriter, err error, allowableError ErrorConceptType, defaultError ErrorConceptType)
```
- Basically, the HTTP status code is swallowed when the following is called: `rr := httptest.NewRecorder()`
- To add to your point, @AnthonyMBonafide, that had led to us creating an anonymous function to be executed when the mock was invoked that would pass the response to the `http.ResponseWriter`, an approach that may not be desirable, as you note, since we are writing the "hard-coded" status code value to the `http.ResponseWriter`. Basically, we achieve that by passing an anonymous function during set up of mocking as a side-effect, something like the following (partially-complete snippet), in outline:
```
func createMockWithVariations(
variations []mockVariations,
w http.ResponseWriter) errorconcept.ErrorHandler {
  ehMock := errorHandlerMock.ErrorHandler{}

  for _, o := range variations {
    ehMock.On(o.methodName, o.args...).Run(func(args mock.Arguments){
      writer := args[0].(http.ResponseWriter)
      writer.WriteHeader()})
  }

  return &ehMock
}
```

Reply:
I agree with @AnthonyMBonafide's suggested approach.